core/src/main/java/org/jruby/RubyObjectSpace.java

```diff
@@ -231,59 +233,74 @@ public WeakMap(Ruby runtime, RubyClass cls) {
 
         @JRubyMethod(name = "[]")
         public IRubyObject op_aref(ThreadContext context, IRubyObject key) {
-            IRubyObject value = map.get(key);
+            Map<IRubyObject, IRubyObject> weakMap = getWeakMapFor(key);
+            IRubyObject value = weakMap.get(key);
             if (value != null) return value;
             return context.nil;
         }
 
+        private Map<IRubyObject, IRubyObject> getWeakMapFor(IRubyObject key) {
+            if (key instanceof RubyFixnum || key instanceof RubyFloat) {
```

Comment:
is it on purpose that we do not handle `RubyInteger`  but only `RubyFixnum` values?

Reply:
Even in CRuby, integer values outside Fixnum range are normal objects.